aips/aip-x.md

```diff
@@ -0,0 +1,237 @@
+
+### Lifecycle of a permissioned handle
+
+The lifecycle of the new signer will look like the following:
+
+```jsx
+module aptos_framework::permissioned_signer {
+    struct PermissionedHandle {
+        master_addr: address,
+        permission_addr: address,
+    }
+
+    public fun create_permissioned_handle(
+        master: &signer,
+    ): PermissionedHandle;
+
+    public fun signer_from_permissioned(p: &PermissionedHandle): signer;
+
+    public fun destroy_permissioned_handle(p: PermissionedHandle);
+
+    public fun is_permissioned_signer(s: &signer): bool;
+}
+```
+
+We would introduce a new struct called `PermissionedHandle` that is derived from the original signer. The handle is generated by `create_permissioned_handle_with_expiration`  API, which takes an original signer. In that handle, we will generate a fresh address using auid API to store the permission information that would be needed by the management API.
+
+Once we have the handle, we can then invoke `signer_from_permissioned`  to obtain a permissioned signer value. The invariant would be that the address of the permissioned signer should be exactly the same as the existing master signer that is used to derive the handle. With the expiration date, this API would abort if the wall clock time has exceeded the expiration date. Also the handle should not be storable. The storable variant will need to follow the logic of account delegation. The handle is also not droppable, so that we can enforce the destruction logic by using `destroy_permissioned_handle`  API.
+
+Note that this handle does not have store capability. There will be storable variant of this handle as well but that would contain an expiry time and the creation of this handle can only be invoked in the Aptos Framework as this is equivalent to signer delegation and thus need special security measure.
+
+### VM Representation of the new signer
+
+Since the signer is not storable, we should have the freedom to change the internal representation of a signer. The new signer will roughly look like the following in the VM:
+
+```rust
+enum Signer {
+    Master(AccountAddress),
+    Permissioned {
+        master_addr: address,
+        permission_addr: address,
+    }
+}
+```
+
+Note that we did change the serialization rules for signer which would be a non-backward compatible change. This should be fine in general as signer should not be storable. The only problem is we exposes `size_of_val`  API in our framework. Since we changed the serialization, this would break the existing semantics for this API. To mitigate this, we will manually make sure that  `size_of_val`  would calculate the size of signer to be a single account address instead of an address plus a variant.
```

Comment:
maybe size_of_val / to_bytes just should abort if signer is inside? (same as we do with delayed fields)

and then let relevant parties know?

because any other solution here is a bit suboptimal

Reply:
I do think we should make it abort. With that said the logic needs to be gated with feature flag.